Add Reverting and Awaiting Bucket Rate Limits

Lakelezz opened the pull request:
There are two new features introduced in this commit:
The first one allows to await a rate limit instead of cancelling the command invocation.
The second one allows to return a ticket to the rate limit bucket.

**Awaiting Rate Limits**
Buckets can now be configured to return an action. The action indicates how a framework should react.
This means, buckets can indicate to cancel or delay command invocation.
In order to set this feature, call the chain-method `await_ratelimits` on the bucket builder.

**Returning Tickets**
If a command never reaches the critical zone that requires a rate limit, we can now undo and return the ticket.
This can be down by returning `RevertBucket` from a command.
The concept is simple, a command may perform an expensive computation or issue a request via web API.
However the command may never reach that part, imagine incorrect user input.
The command would still go on cooldown, which can be a pain for the user.
By returning `RevertBucket`, the cooldown can be undone.

**Bug Fixes**
Following bugs have been discovered and fixed:
- `set_time` being `Option<Instant>` allowed for it being `None` at creation of a new bucket, which allowed twice the ticket count. It's just `Instant` *without* `Option` now.
- If no `limit` is set on the `BucketBuilder`, buckets will always return a rate limit for command invocation. While technically not wrong, logically and ergonomically useless. They now default to the limit of 1.

Please help testing this pull request :)

Lakelezz committed:
Add reverting and awaiting bucket rate limits.

There are two new features introduced in this commit:
The first one allows to await a rate limit instead of cancelling the command invocation.
The second one allows to return a ticket to the rate limit bucket.

**Awaiting Rate Limits**
Buckets can now be configured to return an action. The action indicates how a framework should react.
This means, buckets can indicate to cancel or delay command invocation.
In order to set this feature, call the chain-method `await_ratelimits` on the bucket builder.

**Returning Tickets**
If a command never reaches the critical zone that requires a rate limit, we can now undo and return the ticket.
This can be down by returning `RevertBucket` from a command.
The concept is simple, a command may perform an expensive computation or issue a request via web API.
However the command may never reach that part, imagine incorrect user input.
The command would still go on cooldown, which can be a pain for the user.
By returning `RevertBucket`, the cooldown can be undone.

Flat commented on src/framework/standard/structures/buckets.rs:
```diff
@@ -84,34 +140,98 @@ impl TicketCounter {
+
+        if let Some(ticket_owner) = self.tickets_for.get_mut(&id) {
+
+            // Let's remove a ticket if there is one.
```
```suggestion
            // Remove a ticket if one is available.
```

Flat replied:
```suggestion
            // Remove a ticket if one is available.
```

Flat commented on src/framework/standard/mod.rs:
```diff
@@ -273,15 +275,32 @@ impl StandardFramework {
-        {
-            let mut buckets = self.buckets.lock().await;
+        // Try passing the command's bucket.
+        // We exit the loop if no command ratelimit has been hit or we
```
```suggestion
        // exiting the loop if no command ratelimit has been hit or
```

Flat commented on src/framework/standard/mod.rs:
```diff
@@ -273,15 +275,32 @@ impl StandardFramework {
+        // Try passing the command's bucket.
+        // We exit the loop if no command ratelimit has been hit or we
+        // early-return when ratelimits cancel the framework invocation.
+        // Otherwise, we delay and loop again to check if we passed the bucket.
```
```suggestion
        // Otherwise, delay and loop again to check if we passed the bucket.
```

Flat commented on src/framework/standard/structures/buckets.rs:
```diff
@@ -61,16 +68,65 @@ impl Bucket {
 impl TicketCounter {
-    pub async fn take(&mut self, ctx: &Context, msg: &Message, id: u64) -> Option<Duration> {
+    /// Tries to check whether the invocation is permitted by the ticket counter
+    /// and figuratively speaking if a ticket can be taken; it does not return a
```
```suggestion
    /// and if a ticket can be taken; it does not return a
```

Flat commented on src/framework/standard/structures/buckets.rs:
```diff
@@ -61,16 +68,65 @@ impl Bucket {
+    /// a ticket but a duration until a ticket can be taken.
+    ///
+    /// The duration will be wrapped in an action for the caller to perform
+    /// if wanted. This may inform them to directly cancel trying to take
```
```suggestion
    /// if wanted. This may inform them to directly cancel trying to take a ticket
```

Flat commented on src/framework/standard/structures/buckets.rs:
```diff
@@ -84,34 +140,98 @@ impl TicketCounter {
+            // This tries to bypass a problem of keeping track of when tickets
+            // were taken.
+            // When a ticket is taken, the bucket sets `last_time`, by
+            // substracting the delay once, we allow for one more ticket to be
```
```suggestion
            // substracting the delay, once a ticket is allowed to be
```

Flat commented on src/framework/standard/structures/buckets.rs:
```diff
@@ -84,34 +140,98 @@ impl TicketCounter {
+            // When a ticket is taken, the bucket sets `last_time`, by
+            // substracting the delay once, we allow for one more ticket to be
+            // taken.
+            // If we would reset the value to `None`, we risk resetting the
```
```suggestion
            // If the value is set to `None` this could possibly reset the
```

Flat commented on src/framework/standard/structures/buckets.rs:
```diff
@@ -240,6 +361,18 @@ impl BucketBuilder {
     }
 
+    /// If this is set to `true`, the invocation of the command will be delayed
+    /// and won't return a duration to wait to dispatch erros, but actually
```
```suggestion
    /// and won't return a duration to wait to dispatch errors, but actually
```

Lakelezz committed:
Update src/framework/standard/structures/buckets.rs

Co-authored-by: Ken Swenson <flat@esoteric.moe>

Lakelezz committed:
Update src/framework/standard/mod.rs

Co-authored-by: Ken Swenson <flat@esoteric.moe>

Lakelezz committed:
Update src/framework/standard/structures/buckets.rs

Co-authored-by: Ken Swenson <flat@esoteric.moe>

Lakelezz committed:
Update src/framework/standard/structures/buckets.rs

Co-authored-by: Ken Swenson <flat@esoteric.moe>

Lakelezz committed:
Update src/framework/standard/structures/buckets.rs

Co-authored-by: Ken Swenson <flat@esoteric.moe>

Lakelezz committed:
Update src/framework/standard/structures/buckets.rs

Co-authored-by: Ken Swenson <flat@esoteric.moe>

Lakelezz committed:
Update src/framework/standard/structures/buckets.rs

Co-authored-by: Ken Swenson <flat@esoteric.moe>

Lakelezz committed:
Update src/framework/standard/mod.rs

Co-authored-by: Ken Swenson <flat@esoteric.moe>

Lakelezz committed:
Remove comment duplicate.

Lakelezz committed:
Merge branch 'add-await-duration-to-buckets' of https://github.com/Lakelezz/serenity into add-await-duration-to-buckets

Lakelezz committed:
Fix rate limiting by setting initial `set_time` for buckets.

Lakelezz committed:
Remove dbg!-calls.

Lakelezz committed:
Fix invocation by defaulting ticket limit to 1.

This fixes the bug when no limit is set.
The bucket thought all tickets were exhausted and returned rate limits.
